crates/xtask-bump-check/src/xtask.rs

```diff
@@ -0,0 +1,338 @@
+        };
+
+        insert_changed(delta.old_file().path().unwrap());
+        insert_changed(delta.new_file().path().unwrap());
+    }
+
+    Ok(changed_members)
+}
+
+/// Compares version against published crates on crates.io.
+///
+/// Assumption: We always release a version larger than all existing versions.
+fn check_crates_io<'a>(
+    config: &Config,
+    changed_members: &HashMap<&str, &'a Package>,
+    needs_bump: &mut Vec<&'a Package>,
+) -> CargoResult<()> {
+    let source_id = SourceId::crates_io(config)?;
+    let mut registry = PackageRegistry::new(config)?;
+    let _lock = config.acquire_package_cache_lock()?;
+    registry.lock_patches();
+    config.shell().status(
+        "BumpCheck",
+        format_args!("compare against `{}`", source_id.display_registry_name()),
+    )?;
+    for member in changed_members.values() {
+        let name = member.name();
+        let current = member.version();
+        let version_req = format!("<={current}");
+        let query = Dependency::parse(name, Some(&version_req), source_id)?;
+        let possibilities = loop {
+            // Exact to avoid returning all for path/git
+            match registry.query_vec(&query, QueryKind::Exact) {
+                task::Poll::Ready(res) => {
+                    break res?;
+                }
+                task::Poll::Pending => registry.block_until_ready()?,
+            }
+        };
+        let max_version = possibilities.iter().map(|s| s.version()).max();
+        if max_version >= Some(current) {
+            needs_bump.push(member);
+        }
+    }
+
+    Ok(())
+}
+
+/// Main entry of `xtask-bump-check`.
+///
+/// Assumption: version number are incremental. We never have point release for old versions.
+fn bump_check(args: &clap::ArgMatches, config: &mut cargo::util::Config) -> CargoResult<()> {
+    let ws = args.workspace(config)?;
+    let repo = git2::Repository::open(ws.root()).unwrap();
+    let base_commit = arg_to_commit(args, &repo, "baseline-rev")?;
+    let head_commit = arg_to_commit(args, &repo, "head-rev")?;
+    let referenced_commit = get_referenced_commit(&repo, &base_commit)?;
+    let changed_members = changed(&ws, &repo, &base_commit, &head_commit)?;
+
+    let mut needs_bump = Vec::new();
+
+    check_crates_io(config, &changed_members, &mut needs_bump)?;
+
+    if let Some(commit) = referenced_commit.as_ref() {
+        config.shell().status(
+            "BumpCheck",
+            format_args!("compare against `{}`", commit.id()),
+        )?;
+        for member in checkout_ws(&ws, &repo, commit)?.members() {
+            let name = member.name().as_str();
+            let Some(changed) = changed_members.get(name) else {
+                log::trace!("skpping {name}, may be removed or not published");
+                continue;
+            };
+
+            if changed.version() <= member.version() {
+                needs_bump.push(*changed);
+            }
+        }
```

Comment:
Could we update the variables to clarify these are all reference ones?
- `commit` -> `referenced_commit`
- `member` -> `referenced_member`